spec/importers/envirotech/toolkit_data_spec.rb

```diff
@@ -0,0 +1,65 @@
+
+  before do
+    allow_any_instance_of(Envirotech::IssueData).to receive(:fetch_data).and_return({})
+    allow_any_instance_of(Envirotech::Issue).to receive(:index)
+    allow_any_instance_of(Envirotech::BackgroundLinkData).to receive(:fetch_data).and_return({})
+    allow_any_instance_of(Envirotech::BackgroundLink).to receive(:index)
+    allow_any_instance_of(Envirotech::AnalysisLinkData).to receive(:fetch_data).and_return({})
+    allow_any_instance_of(Envirotech::AnalysisLink).to receive(:index)
+    allow_any_instance_of(Envirotech::ProviderData).to receive(:fetch_data).and_return({})
+    allow_any_instance_of(Envirotech::Provider).to receive(:index)
+    allow_any_instance_of(Envirotech::ProviderSolutionData).to receive(:fetch_data).and_return({})
+    allow_any_instance_of(Envirotech::ProviderSolution).to receive(:index)
+    allow_any_instance_of(Envirotech::RegulationData).to receive(:fetch_data).and_return({})
+    allow_any_instance_of(Envirotech::Regulation).to receive(:index)
+    allow_any_instance_of(Envirotech::SolutionData).to receive(:fetch_data).and_return({})
+    allow_any_instance_of(Envirotech::Solution).to receive(:index)
+
+    scraper = double('Envirotech::ToolkitScraper')
+    allow(Envirotech::ToolkitScraper).to receive(:new).and_return(scraper)
+    allow(scraper).to receive(:all_issue_info).and_return(all_issue_info)
+    allow(JSON).to receive(:parse).and_return(local_data)
+
+    regulation_importer = double('Envirotech::RegulationData')
+    allow(Envirotech::RegulationData).to receive(:new).and_return(regulation_importer)
+    allow(regulation_importer).to receive(:import)
+
+    solution_importer = double('Envirotech::SolutionData')
+    allow(Envirotech::SolutionData).to receive(:new).and_return(solution_importer)
+    allow(solution_importer).to receive(:import)
+  end
+
+  describe '#import' do
+    context 'scraper working as expected' do
+      let(:all_issue_info) { { dummy: 1 } }
+      let(:local_data) { { 'dummy' => 1 } }
+      it 'create importers with relation_data' do
+        expect(Envirotech::RegulationData).to receive(:new).with(relation_data: all_issue_info)
+        expect(Envirotech::SolutionData).to receive(:new).with(relation_data: all_issue_info)
+        importer.import
+      end
+    end
+
+    context 'scraper compairing local data with imported data' do
+      let(:all_issue_info) { { dummy: 1 } }
+      let(:local_data) { { 'dummy' => 2 } }
+      it 'notify Airbrake when data mismatch' do
+        expect(Airbrake).to receive(:notify).with(Exceptions::EnvirotechToolkitDataMismatch.new)
+        importer.import
+      end
+    end
+
+    context 'scraper fallback to local data' do
+      let(:all_issue_info) { }
+      let(:local_data) { { 'dummy' => 2 } }
+      it 'notify Airbrake when UI tool disappear' do
+        expect(Airbrake).to receive(:notify).with(Exceptions::EnvirotechToolkitNotFound.new)
+        importer.import
+      end
+    end
+  end
+end
```

Comment:
Here's what I'm seeing here: `Envirotech::ToolkitData#import` is doing two things. 1. it kicks off all individual Envirotech imports and 2. it figures out if the scraped data is ok (and tells us about it if not). What your spec is testing is a) if `#import` chooses the correct relational data and b) if `Airbrake#notify` is called when it should be. Everything you're testing is related to point 2., and in order to do those tests, you have to mock everything related to 1. This creates a large amount of mocking code, which is a hint that something is not quite right here.

Why don't we break `#import` into two, one part handling the individual imports and the other fetching and checking the relational data? That way, you can write specs for the second method and avoid having to write any mocks at all! In fact, in my enthusiasm I actually started doing that [here](https://github.com/GovWizely/webservices/commit/816a4fdec0c0bb63516e0a0af9cf477bbea4ee20). Can you review my code, and then rework the spec so that it tests the new `fetch_relational_data` method?

In more general language what has happened here is a useful tool (mocking) has got in the way of good object-oriented principles (small, focused methods). It's a subtle trap to fall into, and you should look out for it. The sheer amount of mocking in the spec is the clue that exposes the problem.